src/librustc_codegen_ssa/mir/block.rs

```diff
@@ -21,766 +23,870 @@ use super::place::PlaceRef;
-                                        &bx,
-                                        constant.span,
-                                        constant.ty,
-                                        c,
-                                    );
-                                    return OperandRef {
-                                        val: Immediate(llval),
-                                        layout: bx.layout_of(ty)
-                                    };
-                                }
-                            }
-                        }
+        helper.do_call(self, &mut bx, fn_ty, fn_ptr, &llargs,
+                       destination.as_ref().map(|&(_, target)| (ret_dest, target)),
+                       cleanup);
+    }
+}
 
-                        self.codegen_operand(&mut bx, arg)
-                    }).collect();
+impl<'a, 'tcx: 'a, Bx: BuilderMethods<'a, 'tcx>> FunctionCx<'a, 'tcx, Bx> {
+    pub fn codegen_block(
+        &mut self,
+        bb: mir::BasicBlock,
+    ) {
+        let mut bx = self.build_block(bb);
+        let data = &self.mir[bb];
 
+        debug!("codegen_block({:?}={:?})", bb, data);
 
-                    let callee_ty = instance.as_ref().unwrap().ty(bx.tcx());
-                    bx.codegen_intrinsic_call(callee_ty, &fn_ty, &args, dest,
-                                               terminator.source_info.span);
+        for statement in &data.statements {
+            bx = self.codegen_statement(bx, statement);
+        }
 
-                    if let ReturnDest::IndirectOperand(dst, _) = ret_dest {
-                        self.store_return(&mut bx, ret_dest, &fn_ty.ret, dst.llval);
-                    }
+        self.codegen_terminator(bx, bb, data.terminator());
+    }
 
-                    if let Some((_, target)) = *destination {
-                        funclet_br(self, &mut bx, target);
-                    } else {
-                        bx.unreachable();
-                    }
+    fn codegen_terminator(
+        &mut self,
+        mut bx: Bx,
+        bb: mir::BasicBlock,
+        terminator: &mir::Terminator<'tcx>
+    ) {
+        debug!("codegen_terminator: {:?}", terminator);
 
-                    return;
-                }
+        // Create the cleanup bundle, if needed.
+        let funclet_bb = self.cleanup_kinds[bb].funclet_bb(bb);
+        let helper = TerminatorCodegenHelper {
+            bb: &bb, terminator, funclet_bb
+        };
 
-                // Split the rust-call tupled arguments off.
-                let (first_args, untuple) = if abi == Abi::RustCall && !args.is_empty() {
-                    let (tup, args) = args.split_last().unwrap();
-                    (args, Some(tup))
-                } else {
-                    (&args[..], None)
-                };
+        self.set_debug_loc(&mut bx, terminator.source_info);
+        match terminator.kind {
+            mir::TerminatorKind::Resume => {
+                self.codegen_resume_terminator(helper, bx)
```

Comment:
Pass the helper to the match arm functions when the function would have used one of the closures.